blog/2022/04/simple-chains.md

```diff
@@ -0,0 +1,493 @@
+@def rss_pubdate = Date(2022, 4, 11)
+@def rss = """10x-ing PyTorch: Specializing Scientific Machine Learning with Julia ..."""
+@def published = "11 April 2022"
+@def title = "10x-ing PyTorch: Specializing Scientific Machine Learning with Julia"
```

Comment:
I'm not sure about the 10x claim when the benchmarks in the post does not quite support that.

Reply:
11.24 seconds to 1.324148 seconds?

Reply:
That could be close enough for some "artistic licence" but it also looks a bit cherry-picked as the AMD benchmark is only 5x faster. 

As it stands now, I think the Julia community would not bat an eye but I'd expect a bit of eye-rolling from those feeling a little defensive of PyTorch. 10x-ing is a better punchphrase than something softer or with qualifiers though. @chriselrod would we support our case by adding a more SciML-representative benchmark of a small MLP?

Reply:
We can go with 5x, though we know the benchmark is ultra-conservative given the showcase of how matvecs will have a much larger difference, and the benchmark is matmuls.

Reply:
> 11.24 seconds to 1.324148 seconds?

I'll try a couple things to see if I can squeeze out another 20%.

Reply:
Okay, first thing I tried was about a 2% improvement.

My other idea will require a more radical change, which is transposing the internal data so we can SIMD across batch sizes.

Reply:
Do we have any sort of UDE type example using PyTorch, where we can show we >= 10x them?

It'd be great to have two examples:
1. The intended use case, where hopefully we vastly outperform them.
2. Then we show scaling by moving more towards their domain with the PyTorch example, where we can demonstrate that we were so far ahead in the first, we still have plenty of advantage left over with >= 4x.

Reply:
Unfortunately, I think it'd be difficult to extract much more performance out of this example without larger changes.

SIMD-ing across batch may still be worth it.

Reply:
The problem is that to full specialize the UDE training on SimpleChains, SimpleChains needs to be open sourced so we can integrate it into DiffEqSensitivity.jl and such, which is a chunk of what will be done next 😅.  So I think we just have to choose a stopping point somewhere. If it's easy, try doing the spiral neural ODE (https://gist.github.com/ChrisRackauckas/4a4d526c15cc4170ce37da837bfc32c4) which is from the neural ODE paper.

Reply:
Going full-out SciML and using UDEs in the benchmark might be problematic anyways since much of the time difference could be in the DE solvers. I was more thinking of something like training an MLP to map from a vector of 10 inputs to a single output. That seems like it could be pretty "clean" in the sense that it really only benchmarks the ML tool and representative in that it's the kind of networks we often repeatedly solve in UDEs.